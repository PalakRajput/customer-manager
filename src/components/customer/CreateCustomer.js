import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { createCustomerAction } from '../../redux/actions/CustomerActions';

const CreateCustomer = () => {
    const [firstName, setFirstName] = useState("");
    const [lastName, setLastName] = useState("");
    const [city, setCity] = useState("");
    const [error, setError] = useState("");
    const dispatch = useDispatch();
    const successMessage = useSelector(state => state.customer.successMessage)

    const handleClick = (e) => {
        if ((firstName !== "" && firstName !== null) && (lastName !== "" && lastName !== null) && (city !== "" && city !== null)) {
            setError("")
            dispatch(createCustomerAction({ fname: firstName, lname: lastName, city }))
            setFirstName("");
            setLastName("");
            setCity("");
        } else {
            setError("Please enter data in all fields.")
        }
    }

    return (
        <div className='container' >
            <div className='row' style={{ marginTop: '50px' }}>
                <div className='col-2'>
                    <label htmlFor='fname' style={{ fontWeight: 'bold' }}> First Name: </label>
                    <input data-testid="fname" type="text" id="fname" value={firstName} onChange={(e) => setFirstName(e.target.value)} placeholder="First Name" style={{ height: '40px', borderRadius: '5px' }}></input>
                </div>
                <div className='col-1'></div>
                <div className='col-2'>
                    <label htmlFor='lname' style={{ fontWeight: 'bold' }}> Last Name: </label>
                    <input data-testid="lname" type="text" id="lname" value={lastName} onChange={(e) => setLastName(e.target.value)} placeholder="Last Name" style={{ height: '40px', borderRadius: '5px' }}></input>
                </div>
                <div className='col-1'></div>
                <div className='col-2'>
                    <label htmlFor='city' style={{ fontWeight: 'bold' }}> City: </label>
                    <input data-testid="city" type="text" id="city" value={city} onChange={(e) => setCity(e.target.value)} placeholder="City" style={{ height: '40px', borderRadius: '5px' }}></input>
                </div>
                <div className='col-1'></div>
                <div className='col-2'>
                    <br />
                    <button data-testid="addButton" disabled={!firstName || !lastName || !city} className='btn btn-primary' style={{ width: 'max-content', alignItems: 'center', height: '40px' }} onClick={handleClick}> Add Customer </button>
                </div>
            </div>

            <br />
            {error ? <span className='text-danger font-weight-bold'>{error}</span> : null}
            {/* react-toastify to hide the message after 2secs */}
            {successMessage ? <span className='text-success font-weight-bold'>{successMessage}</span> : null}
        </div>
    )

}

export default CreateCustomer;