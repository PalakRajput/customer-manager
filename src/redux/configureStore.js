import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";
import rootReducer from "./reducers/rootReducer";


const configureStore = createStore(rootReducer, applyMiddleware(thunk));

export default configureStore;

// default exporting the function didn't worked as it gave error in Provider.