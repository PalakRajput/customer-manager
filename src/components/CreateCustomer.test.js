import { render, cleanup, fireEvent, getByTestId, waitFor,    } from '@testing-library/react'
import CreateCustomer from './CreateCustomer'
import React, { Component } from 'react'
import { Provider } from 'react-redux'
import {configureStore} from '../redux/configureStore';
import axiosMock from 'axios';

const store= configureStore

afterEach(cleanup)

describe("Create Customer", () => {
    it (" should render correctly ", () => {
           const {asFragment} =  render(<Provider store={store}><CreateCustomer /></Provider>)
           expect(asFragment(<CreateCustomer />)).toMatchSnapshot();
    })
})

it (" should have initial first name as empty string ", () => {
    const {getByTestId} =  render(<Provider store={store}><CreateCustomer /></Provider>)
    expect(getByTestId('fname')).toHaveDisplayValue("");
})
it (" should have initial last name as empty string ", () => {
    const {getByTestId} =  render(<Provider store={store}><CreateCustomer /></Provider>)
    expect(getByTestId('fname')).toHaveDisplayValue("");
})
it (" should have initial city as empty string ", () => {
    const {getByTestId} =  render(<Provider store={store}><CreateCustomer /></Provider>)
    expect(getByTestId('fname')).toHaveTextContent("");
})
it (" should not have button disabled initially ", () => {
    const {getByTestId} =  render(<Provider store={store}><CreateCustomer /></Provider>)
    expect(getByTestId('addButton')).not.toBeDisabled();
})

it('tests the input text value after onChange event ', () => {
    const { getByTestId } = render(<Provider store={store}><CreateCustomer /></Provider>); 
    fireEvent.change(getByTestId('fname'), {target:{value:'new val'}})
    expect(getByTestId('fname')).toHaveValue('new val')
  });

 jest.mock('axios')

// it('should load and display the data', async () => {
//   const url = 'http://localhost:3000/customres'
//   const { getByTestId, asFragment } = render(<Provider store={store}><CreateCustomer /></Provider>)

//   axiosMock.get("http://localhost:8080/customers").mockResolvedValue({
//     data: { greeting: 'hello there' }, //provide mocked data here
//   })

//   fireEvent.click(getByTestId('fetch-data'))

//   const greetingData = await waitFor(() => getByTestId('show-data'))

//   expect(axiosMock.get).toHaveBeenCalledTimes(1)
//   expect(axiosMock.get).toHaveBeenCalledWith(url)
//   expect(greetingData).toHaveTextContent('hello there')
// })