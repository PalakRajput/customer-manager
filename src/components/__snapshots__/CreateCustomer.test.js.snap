// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Customer  should render correctly  1`] = `
<DocumentFragment>
  <input
    data-testid="fname"
    placeholder="First Name"
    type="text"
    value=""
  />
  <input
    data-testid="lname"
    placeholder="Last Name"
    type="text"
    value=""
  />
  <input
    data-testid="city"
    placeholder="City"
    type="text"
    value=""
  />
  <button
    class="btn btn-primary"
    data-testid="addButton"
  >
     Add Customer 
  </button>
  <br />
</DocumentFragment>
`;
